src/api/array_join.cc

```diff
@@ -0,0 +1,139 @@
+using v8::Object;
+using v8::String;
+
+using iast::tainted::Range;
+
+namespace iast {
+namespace api {
+
+const int DEFAULT_JOIN_SEPARATOR_LENGTH = 1;
+
+void copyRangesWithOffset(Transaction* transaction,
+        SharedRanges* origRanges,
+        SharedRanges* destRanges,
+        int offset) {
+    if (origRanges != nullptr) {
+        auto end = origRanges->end();
+        for (auto it = origRanges->begin(); it != end; it++) {
+            auto origRange = *it;
+            auto newRange = transaction->GetRange(
+                origRange->start + offset,
+                origRange->end + offset,
+                origRange->inputInfo,
+                origRange->secureMarks);
+            if (newRange != nullptr) {
+                destRanges->PushBack(newRange);
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+SharedRanges* getJoinResultRanges(Isolate* isolate,
+        Transaction* transaction, v8::Array* arr,
+        SharedRanges* separatorRanges,
+        int separatorLength) {
+    auto length = arr->Length();
+    int offset = 0;
+    auto newRanges = transaction->GetSharedVectorRange();
+    auto context = isolate->GetCurrentContext();
+    for (uint32_t i = 0; i < length; i++) {
+        if (i > 0) {
+            copyRangesWithOffset(transaction, separatorRanges, newRanges, offset);
+            offset += separatorLength;
+        }
+        auto maybeItem = arr->Get(context, i);
+        if (!maybeItem.IsEmpty()) {
+            auto item = maybeItem.ToLocalChecked();
+            auto taintedItem = transaction->FindTaintedObject(utils::GetLocalPointer(item));
+            auto itemRanges = taintedItem ? taintedItem->getRanges() : nullptr;
+            copyRangesWithOffset(transaction, itemRanges, newRanges, offset);
+            offset += utils::GetLength(isolate, item);
+        }
+    }
+    if (newRanges->Size() == 0) {
+        return nullptr;
+    }
+    return newRanges;
+}
+
+void ArrayJoinOperator(const FunctionCallbackInfo<Value>& args) {
+    auto isolate = args.GetIsolate();
+
+    if (args.Length() < 3) {
+        isolate->ThrowException(v8::Exception::TypeError(
+                        v8::String::NewFromUtf8(isolate,
+                        "Wrong number of arguments",
+                        v8::NewStringType::kNormal).ToLocalChecked()));
+        return;
+    }
+    if (!args[1]->IsString()) {
+        args.GetReturnValue().Set(args[1]);
+        return;
+    }
+
+    auto transaction = GetTransaction(utils::GetLocalPointer(args[0]));
+    if (transaction == nullptr) {
+        args.GetReturnValue().Set(args[1]);
+        return;
+    }
+
+    auto thisArg = args[2];
+    if (thisArg->IsObject()) {
+        auto arrObj = v8::Object::Cast(*args[2]);
+        if (arrObj->IsArray()) {
+            try {
+                int separatorLength = DEFAULT_JOIN_SEPARATOR_LENGTH;
+                SharedRanges* separatorRanges = nullptr;
+                if (args.Length() > 2) {
```

Comment:
if `args.Length() === 3` this if will be `true` and `args[3]` is not going to be defined. `args.Length()` always is > 2, if it isn't, the `if ` in line 83 would be true and exception would be thrown.

```suggestion
                if (args.Length() > 3) {
```

Why is not any test failing with this error?